### ISSUE TITLE
Future pull requests from non-PNNL folks

### ISSUE THREAD

**Issue by cahartin** [JGCRI/hector#92]
...we'll need to be prepared for this w.r.t. IP.

**Comment by rplzzz** [comment]
Turns out there is a relatively easy way to communicate our contribution policies (once we decide on them) to would-be contributors:  https://github.com/blog/1184-contributing-guidelines .  (By the way, note that contribution policies typically cover a lot more than just licenses and copyrights -- we should give some real thought to what we want people to do to validate their contributions before we review them.)

**Comment by bpbond** [comment]
That's an interesting link--thanks. Awesome that GitHub will auto-generate a link when people open a pull request or issue! Definitely a v1.2 issue.

**Comment by bpbond** [comment]
I have created (55744d2935) a `CONTRIBUTING.md` document wrt this issue. @cahartin @rplzzz please take a look.

**Comment by cahartin** [comment]
@bpbond  that's great!

**Comment by rgieseke** [comment]
This will be helpful! Some comments below:

**Getting started**
The link to Hector's version numbering system should probably already here be linked to.

> Trivial changes to comments or documentation do not require creating a new issue. Just start the first line of a commit with '(doc)' instead of an issue number.

Do you want commit messages to start with the issue number? I reference them usually in the description paragraph of the commit message or at the end of the first line. Since this is not used in the commit messages so far it might be helpful to clarify this.
The "proper format" is also referenced in "Making changes", maybe having an example might be useful.

**Making changes**
The branching part is not clear to me yet, maybe just because the branches don't exist yet. Maybe you could elaborate on that. Do you plan to maintain older versions with bug fixes once you switched the major release number?

**Submitting Changes**
I'd think the "statement" requirement and the pull request closing time sound a bit discouraging. The latter I would probably just leave out. Depending on circumstances (and previous interactions) you might want to vary this time span. GitHub's interface makes it also quite easy to re-open something, so you could always do that when the absent-minded professor returns from a sabbatical.

As for the copyright issues, maybe you could word this as a default. The NodeJS project for example has this [paragraph](https://github.com/nodejs/node/blob/master/CONTRIBUTING.md#developers-certificate-of-origin-10):

>By making a contribution to this project, I certify that:
>    (a) The contribution was created in whole or in part by me and I have the right to submit it under the open source license indicated in the file; or
>    (b) The contribution is based upon previous work that, to the best of my knowledge, is covered under an appropriate open source license and I have the right under that license to submit that work with modifications, whether created in whole or in part by me, under the same open source license (unless I am permitted to submit under a different license), as indicated in the file; or
>    (c) The contribution was provided directly to me by some other person who certified (a), (b) or (c) and I have not modified it.

This is probably not perfect for Hector yet, but I would prefer something similar to this. If one had to figure out who in their organization would have to sign off on a small documentation improvement or minor bug fix they might just give up.

Another example, the PISM ice sheet model has their policy at http://www.pism-docs.org/wiki/doku.php?id=committing

I like this part as an explanation on "why" and on how to treat these rules:
>To satisfy our NASA-funded responsibility at UAF, we need to retain some little bit of control. We need PISM to do what we are funded to do. The following list of “best practices” facilitates such minimal control while encouraging healthy development. Perfection on these “rules” is not expected at any time. Mistakes are always forgiven. We are well-protected by version control. 

Hope these ideas make sense ...

**Comment by bpbond** [comment]
Thanks @rgieseke. I've just committed a revised `CONTRIBUTING.md`, incorporating many of your suggestions. One thing I agree with but did not change, yet, is re

>I'd think the "statement" requirement ... sound a bit discouraging...maybe you could word this as a default

@rplzzz what do you think of this suggestion? I definitely agree that asking people to paste in legalese into pull requests is discouraging.

**Comment by rplzzz** [comment]
I agree.  This is actually what I had in mind at the beginning, since CONTRIBUTING.md should be displayed whenever a pull request is opened.  The statement part came in because we didn't have CONTRIBUTING.md in place, and after that it just stayed in through inertia.

The Node.js language looks pretty good.  I think there should be something in there about original authors retaining all other rights.  Not everything people contribute will be derived code covered under the GPL (e.g., new components), and they should be free to use it in non-Hector contexts as they see fit (e.g., if they wanted to contribute it to MAGICC also, to be released under MAGICC's non-open-source license, we should be clear that nothing about contributing to Hector is stopping them from doing that.)

We should also put a last-modified time at the top of the document so that frequent contributors can figure out if anything has changed since the last time they looked at it and skip it if there's nothing new.

**Comment by rgieseke** [comment]
How about changing the licensing paragraph to something like

"Before making a contribution make sure that
- the code was written by you (or the people named in the code)
- you have the right to submit it under Hector's license
- you are authorized to submit it if you're employer or somebody else owns the copyright to the code 

Note that any code written by you can of course be used in other projects under a different license."

I think this would cover everything discussed before and would not sound too legalistic.

Thanks for consideration.

**Comment by swillner** [comment]
I would like to see a more detailled StyleGuide or, even better, a format file to be used for a formatting tool such as [clang-format](https://clang.llvm.org/docs/ClangFormat.html) (styles can be created interactively using https://clangformat.com/). Currently, the code is quite inconsistent regarding formatting.

**Comment by rgieseke** [comment]
There is a [Style Guide](https://github.com/JGCRI/hector/wiki/StyleGuide) on the Wiki, maybe it can be turned into a clang-format description?

**Comment by rplzzz** [comment]
For what it's worth, I'm not a huge fan of elaborate coding style guides, and even less so of reformatting an existing body of code to fit a newly introduced style guide.  On projects that I lead the style guide (if you want to call it that) generally consists of a few simple rules:
1. Use good judgement.
2. Don't write weird code.
3. If it's not obvious what you are trying to do, explain it.
4. When in doubt, talk it over with another developer.

These are based on the observation that elaborate style guides and automatic formatting tools don't seem to prevent people from writing bad code, any more than typesetting guides prevent people from writing bad novels.  

Anyway, that's my advice.  Make of it what you will.

**Comment by bpbond** [comment]
To @rplzzz 's list I would personally add "0. Follow the project's existing style if possible." Some consistency is definitely a good thing IMO. That said, I agree, developing a more-prescriptive style guide is not high on my personal list of priorities, though if someone submits a PR with one, we'd certainly take a look.

But if there are particular spots in Hector's code that need clarifying, please point them out! (We all know a few...) Clarity and concision, in that order, are important to us, and we welcome suggestions for how and where the model code can be improved.

@swillner that's really cool about clang-format--thanks--I didn't know about that.

**Comment by kdorheim** [comment]
@bpbond do we feel that this has been frequently addressed with the addition of the code of conduct, expansion to the style guide, and authorship guidelines that will be included in the package down documentation with  #589?
